core/strings.go

```diff
@@ -0,0 +1,164 @@
+//
+// Package strings implements unicode string abstraction
+// that operates with runes (symbols), not bytes
+//
+// Basic usage:
+//   ustr := NewString("this is unicode string: こんにちは")
+//   slen := ustr.Length()
+//
+//   var str string
+//   str = ustr.String() // equal to original string
+//   usub := ustr.Substring(8, 15) // cut `unicode` word
+//
+//   re, _ := regexp.Compile("string:.*")
+//   matched := ustr.Match(re)  // matched == true
+//   loc := ustr.FindFirst(re)
+//   usub = ustr.Substring(loc[0], loc[1])  // cut `string: こんにちは`
+//
+//   re, _ = regexp.Compile("i")
+//   locs := ustr.FindAll(re)  // locs contain 4 locations (one for every "i" symbol in string)
+//
+//   newS, cnt := ustr.Replace(re, "Ё")  // cnt == 4, newS.String() == "thЁs Ёs unЁcode strЁng: こんにちは""
+//
+
+package strings
+
+import (
+	"regexp"
+)
+
+// Error type that returns ReadRune function
+// when end of string occured
+type noRuneError struct {
+}
+
+// Returns error description for noRuneError
+func (e noRuneError) Error() string {
+	return "no next rune in string"
+}
+
+// Representation of string as an array of runes.
+type String struct {
+	runes []rune
+	ridx  int // used for runes indexing in regexp
+}
+
+// Creates new String object from `string`
+// and returns pointer to this object
+func NewString(orig string) *String {
+	return &String{runes: []rune(orig)}
+}
+
+// Returns length of String object (count of runes)
+func (s *String) Length() int {
+	return len(s.runes)
+}
+
+// Converts String object to `string`
+func (s *String) String() string {
+	return string(s.runes)
+}
+
+// Returns substring as a pointer to new String object
+func (s *String) Substring(from, to int) *String {
+	if from < 0 || from > to || to > len(s.runes) {
+		return nil
+	}
+	runes := s.runes[from:to]
+	new_s := String{runes: runes}
+	return &new_s
+}
+
+// This function is a part of io.RuneReader iterface
+// that used in regexp
+func (s *String) ReadRune() (r rune, size int, err error) {
+	if s.ridx == len(s.runes) {
+		return 0, 0, noRuneError{}
+	}
+	r = s.runes[s.ridx]
+	err = nil
+	s.ridx += 1
+	// padding to 4 bytes per rune
+	// expecting offset returned from regexp.FindReaderIndex equal to rune_offset*4
+	size = 4
+	return
+}
+
+// This internal function convert bytes location to runes location in string
+func (s *String) normalizeRegexpLoc(runes_offset, from, to int) (loc []int) {
+	from_rune := runes_offset + (from / 4)
```

Comment:
1) `rune` is an alias to `int32`.
2) `ReadRune` method always returns size=4.. so we expecting offset returned from regexp.FindReaderIndex equal to rune_offset*4